File: src/go/bots/heartbeat/heartbeat.go
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"flag"
	"fmt"
	"os"
	"path"
	"strconv"
	"strings"
	"time"

	"cloud.google.com/go/datastore"
	"github.com/shirou/gopsutil/process"

	"clusterfuzz/go/base/logs"
	"clusterfuzz/go/base/persist"
	baseProc "clusterfuzz/go/base/process"
	"clusterfuzz/go/bots"
	"clusterfuzz/go/bots/tasks"
	"clusterfuzz/go/cloud/db"
	"clusterfuzz/go/cloud/db/types"
	"clusterfuzz/go/cloud/stackdriver"
)

const (
	heartbeatWaitInterval = 10 * time.Minute
	heartbeatUpdateKey    = "heartbeat_update"
)

// updateHeartbeat updates the heartbeat entity.
func updateHeartbeat() error {
	now := time.Now()

	// Check if heartbeat was recently updated and bail out if so.
	if persist.HasValue(heartbeatUpdateKey) {
		lastModTime := time.Unix(persist.GetValueInt(heartbeatUpdateKey), 0)
		if now.Before(lastModTime.Add(heartbeatWaitInterval)) {
			return nil
		}
	}

	key := datastore.NameKey("Heartbeat", bots.BotName(), nil)

	ctx := context.Background()
	var heartbeat types.Heartbeat

	err := db.Get(ctx, key, &heartbeat)
	if err != nil && !db.IsNoSuchEntityError(err) {
		return err
	}

	heartbeat.BotName = bots.BotName()
	heartbeat.TaskPayload = tasks.CurrentTaskPayload()
	heartbeat.TaskEndTime = tasks.CurrentTaskEndTime()
	heartbeat.LastBeatTime = now
	heartbeat.SourceVersion = bots.CurrentSourceVersion()

	_, err = db.Put(ctx, key, &heartbeat)
	if err != nil {
		return err
	}

	persist.SetValue(heartbeatUpdateKey, now.Unix())
	return nil
}

// killStuckTask kills the current bot instance if the task is stuck.
func killStuckTask() error {
	if !tasks.IsTaskRunning() {
		return nil
	}

	taskEndTime := tasks.CurrentTaskEndTime()
	if time.Now().Before(taskEndTime.Add(tasks.TaskCompletionBuffer)) {
		return nil
	}

	botFilePath := path.Join(bots.RootDir(), "src", "bot", "run_bot")
	procs, err := process.Processes()
	if err != nil {
		return err
	}

	for _, proc := range procs {
		cmdline, err := proc.Cmdline()
		if err != nil {
			logs.Warnf("Failed to get cmdline for process: %v", err)
			continue
		}

		if strings.Contains(cmdline, botFilePath) {
			logs.Logf("Killing stale bot (pid=%d, cmd=%s) which seems to be stuck.", proc.Pid, cmdline)
			err = baseProc.KillTree(proc.Pid)
			if err != nil {
				logs.Errorf("Failed to kill process %d: %+v", proc.Pid, err)
			}
		}
	}

	tasks.TrackTaskEnd()
	err = bots.ClearTempDirectories(true)
	if err != nil {
		return err
	}

	return bots.ClearTestcaseDirectories()
}

// beat does a heartbeat.
func beat(prevState int64, logFilename string) (int64, error) {
	logs.Logf("Beat for logging pipeline.")

	err := killStuckTask()
	if err != nil {
		return 0, err
	}

	logStat, err := os.Stat(logFilename)
	if err != nil {
		return 0, err
	}

	curState := logStat.ModTime().Unix()
	logs.Logf("Old state %d, new state %d.", prevState, curState)
	if curState != prevState {
		logs.Logf("State updated.")
		return curState, updateHeartbeat()
	}

	return curState, nil
}

func main() {
	flag.Parse()
	if flag.NArg() != 2 {
		fmt.Fprintf(flag.CommandLine.Output(), "Usage: %s prev_state log_path", os.Args[0])
		os.Exit(1)
	}

	bots.SetUpEnvironment()
	db.Init()

	logPath := path.Join(bots.LogDir(), "heartbeat_go.log")
	logger, err := stackdriver.Create(logPath)
	if err != nil {
		logs.Panicf("Failed to create logger: %+v", err)
	}
	logs.Init(logger)

	prevState, err := strconv.ParseInt(flag.Arg(0), 10, 64)
	if err != nil {
		logs.Panicf("Failed to parse previous state %s: %v.", flag.Arg(0), err)
	}

	logFilename := flag.Arg(1)
	result, err := beat(prevState, logFilename)

	if err != nil {
		logs.Errorf("Failed to beat: %+v.", err)
	} else {
		fmt.Printf("%d", result)
	}
	time.Sleep(heartbeatWaitInterval)
}

File: src/python/bot/startup/heartbeat.py
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Heartbeat script that monitors
   whether the bot is still running or not."""

# Before any other imports, we must fix the path. Some libraries might expect
# to be able to import dependencies directly, but we must store these in
# subdirectories of common so that they are shared with App Engine.
from python.base import modules
modules.fix_module_search_paths()

import os
import sys
import time

from base import dates
from base import tasks
from datastore import data_handler
from datastore import data_types
from datastore import ndb_init
from metrics import logs
from system import environment
from system import process_handler
from system import shell

try:
  import psutil
except ImportError:
  psutil = None


def beat(previous_state, log_filename):
  """Run a cycle of heartbeat checks to ensure bot is running."""
  # Handle case when run_bot.py script is stuck. If yes, kill its process.
  task_end_time = tasks.get_task_end_time()
  if psutil and task_end_time and dates.time_has_expired(
      task_end_time, seconds=tasks.TASK_COMPLETION_BUFFER):

    # Get absolute path to |run_bot| script. We use this to identify unique
    # instances of bot running on a particular host.
    startup_scripts_directory = environment.get_startup_scripts_directory()
    bot_file_path = os.path.join(startup_scripts_directory, 'run_bot')

    for process in psutil.process_iter():
      try:
        command_line = ' '.join(process.cmdline())
      except (psutil.AccessDenied, psutil.NoSuchProcess, OSError):
        continue

      # Find the process running the main bot script.
      if bot_file_path not in command_line:
        continue

      process_id = process.pid
      logs.log(
          'Killing stale bot (pid %d) which seems to have stuck.' % process_id)
      try:
        process_handler.terminate_root_and_child_processes(process_id)
      except Exception:
        logs.log_error('Failed to terminate stale bot processes.')

    # Minor cleanup to avoid disk space issues on bot restart.
    process_handler.terminate_stale_application_instances()
    shell.clear_temp_directory()
    shell.clear_testcase_directories()

    # Concerned stale processes should be killed. Now, delete the stale task.
    tasks.track_task_end()

  # Figure out when the log file was last modified.
  try:
    current_state = str(os.path.getmtime(log_filename))
  except Exception:
    current_state = None

  # Only update the heartbeat if the log file was modified.
  if current_state and current_state != previous_state:
    # Try updating the heartbeat. If an error occurs, just
    # wait and return None.
    if not data_handler.update_heartbeat():
      return None
    # Heartbeat is successfully updated.

  return current_state


def main():
  logs.configure('heartbeat')
  dates.initialize_timezone_from_environment()
  environment.set_bot_environment()

  if sys.argv[1] == 'None':
    previous_state = None
  else:
    previous_state = sys.argv[1]

  log_filename = sys.argv[2]

  try:
    sys.stdout.write(str(beat(previous_state, log_filename)))
  except Exception:
    logs.log_error('Failed to beat.')

  time.sleep(data_types.HEARTBEAT_WAIT_INTERVAL)


if __name__ == '__main__':
  with ndb_init.context():
    main()
